### PR TITLE
[SystemC] Add SignalOp

### PR DESCRIPTION
**Pull request by maerhart**
Limits the allowed types for signals to signless integers for now. We can expand that later to all the possible types allowed according to IEEE 1666-2011 §6.4.3. as needed.

Aims to implement names not as hints but guarantees that the name given to a signal, port, etc. during lowering to SystemC is preserved up to emission. Name uniqueness checking is currently implemented as verification on SCModuleOp, but if it turns out to be too expensive we can move it to a separate pass that can then be run at positions in the pipline where it makes sense. However, as the systemc dialect is only for emission, there shouldn't be a lot of transformations on it and thus not a lot of verification runs.

### CODE REVIEW

**Comment by seldridge** [comment]
> but if it turns out to be too expensive we can move it to a separate pass that can then be run at positions in the pipline where it makes sense

I would advise against ever making a verification check something that runs in a pass. It is reasonable to change this to an `ExportVerilog`-style pre-pass that uniquifies names, however.

File: include/circt/Dialect/SystemC/SystemCOpInterfaces.td [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// This describes the SystemC operation interfaces.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_SYSTEMC_SYSTEMCOPINTERFACES
+#define CIRCT_DIALECT_SYSTEMC_SYSTEMCOPINTERFACES
+
+include "mlir/IR/OpBase.td"
+
+def SystemCNameDeclOp : OpInterface<"SystemCNameDeclOp"> {
```
Comment (seldridge) [inline include/circt/Dialect/SystemC/SystemCOpInterfaces.td:18]: This is actually defining an `OpInterface`, yes? Is there a convention for what this should be called as opposed to calling it `SystemCNameDeclOp`? This implies to me that it is an op, when it is actually an opinterface.

File: test/Dialect/SystemC/module-error.mlir [hunk 2]
```diff
@@ -35,5 +35,50 @@
...
+  %0 = "systemc.signal"() {name = "signal0"} : () -> i32
+  %1 = "systemc.signal"() {name = "signal0"} : () -> i32
+}
+
+// -----
+
+// expected-error @+1 {{declared names must be unique}}
+systemc.module @signalNameConflictWithArg (sc_in %in: i32) {
+  %0 = "systemc.signal"() {name = "in"} : () -> i32
+}
```
Comment (seldridge) [inline test/Dialect/SystemC/module-error.mlir:61]: This type of test would motivate additional error reporting that the signal name conflicts with the port name. The general form of this is probably changing the reporting to state: 
1. An error on the SCModule (not printing the full module).
2. Iterating over all the conflicts and printing the problematic ops/ports.

File: lib/Dialect/SystemC/SystemCOps.cpp [hunk 3]
```diff
@@ -173,6 +186,51 @@ LogicalResult SCModuleOp::verify() {
   return success();
 }
 
+LogicalResult SCModuleOp::verifyRegions() {
+  DenseSet<StringRef> memberNames;
+  DenseSet<StringRef> localNames;
```
Comment (seldridge) [inline lib/Dialect/SystemC/SystemCOps.cpp:191]: To facilitate better error reporting where the list of all conflicting ops are printed, these could be changed to use `llvm::EquivalenceClasses` (a disjoint set) or `DenseMap<StringRef, SmallVector<Operation *>>`. This is likely overkill and it is sufficient to just print out the ops that conflict as soon as they are found.

**Review by seldridge — APPROVED** [review]
Looks good. 

There are some suggestions on improving the error reporting that likely require including more information during the walk. Not critical.

File: include/circt/Dialect/SystemC/SystemCOpInterfaces.td (hunk 1 above)
Comment (maerhart) [reply include/circt/Dialect/SystemC/SystemCOpInterfaces.td:18]: You're right, the default seems to be to have `OpInterface` at the end, so I'll add the 'Interface'.

File: test/Dialect/SystemC/module-error.mlir (hunk 2 above)
Comment (maerhart) [reply test/Dialect/SystemC/module-error.mlir]: I changed your suggestion slightly to report the error on the op that redefines the name and add notes to the first definition of the name as well as the module the conflict occurs in.

**Pull request merged**